### PR TITLE
Volume API: Add action, metadata and snapshot metadata methods

### PR DESCRIPTION
**Pull request by dhague**

### CODE REVIEW

**Comment by gildub** [comment]
LGTM.

Makes sense to wait after #82 (see comments).

File: lib/fog/openstack/models/volume_v2/snapshot.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+          attribute :name
+          attribute :status
+          attribute :description
+          attribute :metadata
+          attribute :force
+
+          def save
+            requires :name
+            if self.id.nil?
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v2/snapshot.rb]: I'm not sure self is required here. Same for following ones below.

File: lib/fog/openstack/models/volume_v1/volume.rb [hunk 2]
```diff
@@ -14,11 +14,21 @@ class Volume < Fog::Volume::OpenStack::Volume
 
           def save
             requires :display_name, :size
-            data = service.create_volume(display_name, display_description, size, attributes)
+            if self.id.nil?
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/volume.rb]: I'm not sure self is required here. Same for following ones below.

File: lib/fog/openstack/volume_v1.rb [hunk 3]
```diff
@@ -32,22 +35,24 @@ class V1 < Fog::Volume::OpenStack
...
         request :create_volume
+        request :update_volume
         request :get_volume_details
         request :extend_volume
         request :delete_volume
 
         request :list_zones
 
         request :list_volume_types
-        request :create_volume_type
```
Comment (gildub) [inline lib/fog/openstack/volume_v1.rb]: What's happening with this request? Why is it disappearing?
Same for following one.

File: lib/fog/openstack/volume_v2.rb [hunk 4]
```diff
@@ -61,9 +68,12 @@ class V2 < Fog::Volume::OpenStack
         request :get_quota
         request :get_quota_defaults
 
-        request :get_quota_usage
```
Comment (gildub) [inline lib/fog/openstack/volume_v2.rb]: Same question here, where is that request going?

**Comment by gildub** [comment]
There are some requests that seems to have disappeared (files lib/fog/openstack/volume_v1.rb and volume_v2.rb too, please see comments)
Not sure why, the corresponding file are still there, maybe I missed something. But if not then don't we need a depreciation approach?

Please fix hounds including replace simple quotes by double quotes. I use to do the same but single ones seems out of trend now, well according to Rubocop!

Depends upon #82 too, (I'll review the spec after merge).

**Comment by dhague** [comment]
Missing requests are reinstated - thanks for catching hat, @gildub, looks like they somehow got lost as I was merging.
I don't see any need for adding to examples, because the volume_spec.rb tests serve exactly the same purpose - each test shows how to call the related API call.
I'll have a quick scan through the outstanding Hound issues, and will address any related to coding style. I do not plan to address any of the ones related to whitespace, because life is just too short...

File: lib/fog/openstack/models/volume_v1/volume.rb [hunk 5]
```diff
@@ -14,11 +14,21 @@ class Volume < Fog::Volume::OpenStack::Volume
 
           def save
             requires :display_name, :size
-            data = service.create_volume(display_name, display_description, size, attributes)
+            if id.nil?
+              data = service.create_volume(display_name, display_description, size, attributes)
+            else
+              data = service.update_volume(self.id, attributes.reject{|k,v| k == :id})
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/volume.rb]: No need to use `self.id`, just `id` as earlier.

File: lib/fog/openstack/models/volume_v1/volume.rb [hunk 6]
```diff
@@ -14,11 +14,21 @@ class Volume < Fog::Volume::OpenStack::Volume
...
+            end
             merge_attributes(data.body['volume'])
             true
           end
 
+          def update(attr = nil)
+            requires :id
+            merge_attributes(
+                service.update_volume(self.id, attr || attributes).body['volume'])
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/volume.rb]: Same, no need to use `self`.

File: lib/fog/openstack/models/volume_v2/snapshot.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+          attribute :name
+          attribute :status
+          attribute :description
+          attribute :metadata
+          attribute :force
+
+          def save
+            requires :name
+            if id.nil?
+              data = service.create_snapshot(self.attributes[:volume_id], name, description, force)
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v2/snapshot.rb]: Same, no need to use self.

File: lib/fog/openstack/models/volume_v2/snapshot.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+          end
+
+          def create
+            requires :name
+
+            #volume_id, name, description, force=false
+            response = service.create_snapshot(self.attributes[:volume_id],
+                                                      self.attributes[:name],
+                                                      self.attributes[:description],
+                                                      self.attributes[:force])
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v2/snapshot.rb]: All 4 lines above, no need to use self.

File: lib/fog/openstack/models/volume_v2/snapshot.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+          attribute :description
+          attribute :metadata
+          attribute :force
+
+          def save
+            requires :name
+            if id.nil?
+              data = service.create_snapshot(self.attributes[:volume_id], name, description, force)
+            else
+              data = service.update_snapshot(self.id, attributes.reject{|k,v| k == :id})
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v2/snapshot.rb]: Same, no need to use self.

File: lib/fog/openstack/models/volume_v1/snapshot.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+          end
+
+          def create
+            requires :display_name
+
+            #volume_id, name, description, force=false
+            response = service.create_snapshot(self.attributes[:volume_id],
+                                                      self.attributes[:display_name],
+                                                      self.attributes[:display_description],
+                                                      self.attributes[:force])
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/snapshot.rb]: All 4 lines above, no need to use self.

File: lib/fog/openstack/models/volume/snapshot.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+module Fog
+  module Volume
+    class OpenStack
+      class Snapshot < Fog::OpenStack::Model
+
+        def update(data)
+          requires :id
+
+          response = service.update_snapshot(self.id, data)
```
Comment (gildub) [inline lib/fog/openstack/models/volume/snapshot.rb]: No need to use `self.id`, just `id`.

File: lib/fog/openstack/models/volume/snapshot.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+          merge_attributes(response.body['snapshot'])
+
+          self
+        end
+
+        def destroy
+          requires :id
+          service.delete_snapshot(self.id)
```
Comment (gildub) [inline lib/fog/openstack/models/volume/snapshot.rb]: Same as above, no need to use `self.`.

File: lib/fog/openstack/models/volume_v1/snapshot.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+          attribute :display_name
+          attribute :status
+          attribute :display_description
+          attribute :metadata
+          attribute :force
+
+          def save
+            requires :display_name
+            if id.nil?
+              data = service.create_snapshot(self.attributes[:volume_id], display_name, display_description, force)
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/snapshot.rb]: Same as above, no need to use self..

File: lib/fog/openstack/models/volume_v1/snapshot.rb [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            else
+              data = service.update_snapshot(self.id, attributes.reject{|k,v| k == :id})
```
Comment (gildub) [inline lib/fog/openstack/models/volume_v1/snapshot.rb]: Same as above, no need to use self..

File: lib/fog/openstack/models/volume/snapshot.rb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        def delete_metadata(key_name)
+          requires :id
+          service.delete_snapshot_metadata(id, key_name)
+          true
+        end
+
+      end
+    end
+  end
+end
```
Comment (gildub) [inline lib/fog/openstack/models/volume/snapshot.rb]: Seems like a end of file is missing.

File: lib/fog/openstack/models/volume/snapshot.rb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+          true
+        end
+
+        # Delete one specific key-value pair by specifying the key name
+        def delete_metadata(key_name)
+          requires :id
+          service.delete_snapshot_metadata(id, key_name)
+          true
+        end
+
```
Comment (gildub) [inline lib/fog/openstack/models/volume/snapshot.rb]: Please remove empty line (sorry for such cosmetic and annoying comment!).

File: lib/fog/openstack/models/volume/snapshots.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+        end
+
+        def get(snapshots_id)
+          if snapshot = service.get_snapshot_details(snapshots_id).body['snapshot']
+            new(snapshot)
+          end
+        rescue Fog::Volume::OpenStack::NotFound
+          nil
+        end
+
```
Comment (gildub) [inline lib/fog/openstack/models/volume/snapshots.rb]: Same empty line cosmetic comment. Just to be consistent, most of the time it's not there. I know most of the legacy stuff is not aligned too :(

File: lib/fog/openstack/volume_v1.rb [hunk 18]
```diff
@@ -43,11 +47,14 @@ class V1 < Fog::Volume::OpenStack
         request :delete_volume_type
         request :get_volume_type_details
 
-        request :create_volume_snapshot
```
Comment (gildub) [inline lib/fog/openstack/volume_v1.rb]: I believe this request is still needed and shouldn't be removed.

File: lib/fog/openstack/volume_v2.rb [hunk 19]
```diff
@@ -44,11 +48,14 @@ class V2 < Fog::Volume::OpenStack
         request :delete_volume_type
         request :get_volume_type_details
 
-        request :create_volume_snapshot
```
Comment (gildub) [inline lib/fog/openstack/volume_v2.rb]: I believe this request is still needed and shouldn't be removed.

**Comment by gildub** [comment]
Looks good beside another missing request which seems to be missing again.
Per my comments there are few usage of self where it's not needed.
I think a pass is needed to replace single quote with double.

File: lib/fog/openstack/volume_v1.rb (hunk 18 above)
Comment (dhague) [inline lib/fog/openstack/volume_v1.rb]: In this case the request files named "create_volume_snapshot.rb" should be deleted. "create_snapshot.rb" is used instead, which matches the pattern of the existing delete_snapshot.rb, update_snapshot.rb and list_snapshots.rb files.

**Comment by dhague** [comment]
All instances of "self.*" in the Volume API are now corrected, and the missing request is removed entirely (as per my comment above, create_volume_snapshot was a duplicate of create_snapshot).
I will not be going through replacing single quotes with doubles for this PR. We should discuss with the team in the Gitter chat what we think about the quotes issue and how best to tackle it (possibly a single over-arching PR to enforce style, or maybe relaxing the style guidelines).

**Comment by gildub** [comment]
Thanks for fixing it and removing the create_volume_snapshot files.

**Comment by gildub** [comment]
Still some hounds to fix or change the rules

File: lib/fog/openstack/requests/volume/replace_metadata.rb [hunk 20]
```diff
@@ -0,0 +1,20 @@
+module Fog
+  module Volume
+    class OpenStack
+      module Real
+        def replace_metadata(volume_id, metadata={})
```
Comment (gildub) [inline lib/fog/openstack/requests/volume/replace_metadata.rb]: What about this one?

File: lib/fog/openstack/requests/volume/replace_metadata.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+          data = {
+              'metadata' => metadata
```
Comment (gildub) [inline lib/fog/openstack/requests/volume/replace_metadata.rb]: One here too.

File: lib/fog/openstack/requests/volume/replace_metadata.rb [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+          }
+
+          request(
+              :body    => Fog::JSON.encode(data),
```
Comment (gildub) [inline lib/fog/openstack/requests/volume/replace_metadata.rb]: Looks like you editor is setup with 4 indent by default.

File: lib/fog/openstack/requests/volume/delete_snapshot_metadata.rb [hunk 23]
```diff
@@ -0,0 +1,15 @@
+module Fog
+  module Volume
+    class OpenStack
+      module Real
+        def delete_snapshot_metadata(snapshot_id, key_name)
+          request(
+              :expects => [200],
```
Comment (gildub) [inline lib/fog/openstack/requests/volume/delete_snapshot_metadata.rb]: Same indent issue

File: lib/fog/openstack/requests/volume/delete_metadata.rb [hunk 24]
```diff
@@ -0,0 +1,15 @@
+module Fog
+  module Volume
+    class OpenStack
+      module Real
+        def delete_metadata(volume_id, key_name)
+          request(
+              :expects => [200],
```
Comment (gildub) [inline lib/fog/openstack/requests/volume/delete_metadata.rb]: Indent 4 too.

**Comment by gildub** [comment]
Almost there just few indents of 4 instead of 2.

**Comment by dhague** [comment]
Where do you see the indent issues? Hound is reporting no violations...

**Comment by gildub** [comment]
I see them when I look in the "Files changed" tab, which I confirmed by cloning the PR locally to double check the files.

**Comment by dhague** [comment]
Sorry, maybe I am being blind but I just don't see them. When I go to the "Files changed" tab in GitHub and search for the word "indent", nothing is found.
Could you tell me specifically which files & line numbers to look at?

**Comment by gildub** [comment]
Well I don't see them anymore, weird, sorry about the trouble, maybe I worked on version between commits. Anyway LGTM now. Thanks

**Pull request merged**